src/JuliaSyntaxHighlighting.jl

```diff
@@ -59,96 +69,232 @@ const HIGHLIGHT_FACES = [
-        end
+function paren_type(k::Kind)
+    if     k == K"(";  1, :paren
+    elseif k == K")"; -1, :paren
+    elseif k == K"[";  1, :bracket
+    elseif k == K"]"; -1, :bracket
+    elseif k == K"{";  1, :curly
+    elseif k == K"}"; -1, :curly
+    else               0, :none
     end
-    depthcounters = (paren = Ref(0), bracket = Ref(0), curly = Ref(0))
-    for (; head::JuliaSyntax.SyntaxHead, range::UnitRange{UInt32}) in tokens
-        range = first(range):thisind(content, last(range))
-        kind = head.kind
-        face = if kind == K"Identifier"
-            if lastk == K":" && !JuliaSyntax.is_number(last2k) &&
-                last2k ∉ (K"Identifier", K")", K"]", K"end", K"'")
-                highlighted[end] = (highlighted[end][1], :face => :julia_symbol)
-                :julia_symbol
-            elseif lastk == K"::"
-                :julia_type
-            elseif lastk ∈ (K".", K"{") && last2f == :julia_type
-                :julia_type
-            elseif view(content, range) in SINGLETON_IDENTIFIERS
+end
+
+struct ParenDepthCounter
+    paren::Ref{UInt}
+    bracket::Ref{UInt}
+    curly::Ref{UInt}
+end
+
+ParenDepthCounter() =
+    ParenDepthCounter(Ref(zero(UInt)), Ref(zero(UInt)), Ref(zero(UInt)))
+
+struct GreenLineage
+    node::GreenNode
+    parent::Union{Nothing, GreenLineage}
+end
+
+struct HighlightContext{S <: AbstractString}
+    content::S
+    offset::Int
+    lnode::GreenNode
+    pdepths::ParenDepthCounter
+end
+
+function _hl_annotations(content::AbstractString, ast::GreenNode)
+    highlights = Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}()
+    ctx = HighlightContext(content, 0, ast, ParenDepthCounter())
+    _hl_annotations!(highlights, GreenLineage(ast, nothing), ctx)
+    highlights
+end
+
+function _hl_annotations!(highlights::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}},
+                          lineage::GreenLineage, ctx::HighlightContext)
+    (; node, parent) = lineage
+    (; content, offset, lnode, pdepths) = ctx
+    region = firstindex(content)+offset:node.span+offset
+    nkind = node.head.kind
+    pnode = if !isnothing(parent) parent.node end
+    pkind = if !isnothing(parent) kind(parent.node) end
+    ppkind = if !isnothing(parent) && !isnothing(parent.parent)
+        kind(parent.parent.node) end
+    # We need to check for infix binary operators
+    # of the form [lhs] [ws] [op] [ws]? [rhs]
+    isbinaryinfix(node) = 3 <= length(node.args) <= 5 &&
+        (JuliaSyntax.is_operator(kind(node.args[2])) ||
+        (kind(node.args[2]) == K"Whitespace" &&
+        JuliaSyntax.is_operator(kind(node.args[3]))))
+    isplainoperator(node) =
+        JuliaSyntax.is_operator(node) &&
+        !JuliaSyntax.is_trivia(node) &&
+        !JuliaSyntax.is_prec_assignment(node) &&
+        !JuliaSyntax.is_word_operator(node) &&
+        nkind != K"." && nkind != K"..." &&
+        (JuliaSyntax.is_trivia(node) ||
+        iszero(flags(node)) && !isbinaryinfix(node))
```

Comment:
Are you looking for `!JuliaSyntax.haschildren(node)` rather than testing `isbinaryinfix()`?

What is `iszero(flags(node))` supposed to be doing?

Reply:
I think just `!JuliaSyntax.haschildren(node)` might do the trick, I'll give it a whirl and see :)